presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java

```
@@ -0,0 +1,227 @@
+
+        comparisons.forEach(System.out::println);
+    }
+
+    private List<MetricComparison> getMetricComparisons(QueryPlan queryPlan, QueryStats queryStats)
+    {
+        return METRICS.entrySet().stream().flatMap(metricEntry -> {
+            String metricName = metricEntry.getKey();
+            Function<PlanNodeCost, Estimate> metricFunction = metricEntry.getValue();
+            Map<PlanNode, PlanNodeCost> estimatesByPlanNode = getCostEstimatesByPlanNode(queryPlan);
+            return estimatesByPlanNode.entrySet().stream().map(nodeCostEntry -> {
+                PlanNode node = nodeCostEntry.getKey();
+                PlanNodeCost estimate = nodeCostEntry.getValue();
+                Map<PlanNodeId, PlanNodeCost> executionCostsByPlanId = getExecutionCostsById(queryStats);
+                Optional<PlanNodeCost> execution = Optional.ofNullable(executionCostsByPlanId.get(node.getId()));
+                Optional<Double> estimatedCost = asOptional(metricFunction.apply(estimate));
+                Optional<Double> executionCost = execution.flatMap(e -> asOptional(metricFunction.apply(e)));
+                return new MetricComparison(node, metricName, estimatedCost, executionCost);
+            });
+        }).collect(Collectors.toList());
+    }
+
+    private Map<PlanNode, PlanNodeCost> getCostEstimatesByPlanNode(QueryPlan queryPlan)
+    {
+        return queryPlan.getPlanNodeCosts().entrySet().stream()
+                .collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+    }
+
+    private Map<PlanNodeId, PlanNodeCost> getExecutionCostsById(QueryStats queryStats)
+    {
+        //for some reason, there are duplicate PlanNodeId-s in operatorSummaries.
+        BinaryOperator<PlanNodeCost> mergeOnlyIfEqual = (a, b) -> {
+            assertEquals(a, b);
+            return b;
+        };
+        return queryStats.getOperatorSummaries().stream()
+                .collect(toMap(OperatorStats::getPlanNodeId, this::toPlanNodeCost, mergeOnlyIfEqual));
```

What happens when costs are different. Does the pipeline quietly fail?

---

It fails loudly with an `AssertionError`